src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,84 +1081,108 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
-                                 "[type={}], CRC mismatch: {}/{}, on mblk bid: {}, context_sz: {}", mblk->hdr.h.type,
-                                 crc, uint32_cast(mblk->hdr.h.crc), mblk->hdr.h.bid.to_string(),
-                                 uint64_cast(mblk->hdr.h.context_sz));
-            } else {
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk found with bypassing crc.", mblk->hdr.h.type);
-            }
 
-            // send the callbck;
-            auto& cb = itr->second.cb;
-            if (cb) { // cb could be nullptr because client want to get its superblock via read api;
-                // decompress if necessary
-                if (mblk->hdr.h.compressed) {
-                    // HS_DBG_ASSERT_GE(mblk->hdr.h.context_sz, META_BLK_CONTEXT_SZ);
-                    // TO DO: Might need to address alignment based on data or fast type
-                    auto decompressed_buf{hs_utils::make_byte_array(mblk->hdr.h.src_context_sz, true /* aligned */,
-                                                                    sisl::buftag::compression, align_size())};
-                    size_t decompressed_size = mblk->hdr.h.src_context_sz;
-                    const auto ret{sisl::Compress::decompress(r_cast< const char* >(buf->bytes),
-                                                              r_cast< char* >(decompressed_buf->bytes),
-                                                              mblk->hdr.h.compressed_sz, &decompressed_size)};
-                    if (ret != 0) {
-                        LOGERROR("[type={}], negative result: {} from decompress trying to decompress the "
-                                 "data. compressed_sz: {}, src_context_sz: {}",
-                                 mblk->hdr.h.type, ret, uint64_cast(mblk->hdr.h.compressed_sz),
-                                 uint64_cast(mblk->hdr.h.src_context_sz));
-                        HS_REL_ASSERT(false, "failed to decompress");
-                    } else {
-                        // decompressed_size must equal to input sz before compress
-                        HS_REL_ASSERT_EQ(uint64_cast(mblk->hdr.h.src_context_sz),
-                                         uint64_cast(decompressed_size)); /* since decompressed_size is >=0 it
-                                                                             is safe to cast to uint64_t */
-                        HS_LOG(DEBUG, metablk,
-                               "[type={}] Successfully decompressed, compressed_sz: {}, src_context_sz: {}, "
-                               "decompressed_size: {}",
-                               mblk->hdr.h.type, uint64_cast(mblk->hdr.h.compressed_sz),
-                               uint64_cast(mblk->hdr.h.src_context_sz), decompressed_size);
-                    }
+    meta_subtype_vec_t independent_subtypes;
+    meta_subtype_vec_t ordered_subtypes;
+
+    if (hs_utils::topological_sort(m_dep_topo_graph, ordered_subtypes, &independent_subtypes)) {
```

Comment:
Can we move this sort out of MetaBlkService::recover and put it into either MetaBlkService::start (before ::recover) or into a new function (e.g. validate_register_dep(...)) to indicate this is a validation of register dependency, since this is not part of recovery.

Reply:
as i mentioned, no matter it is a fresh boot or not, the order is `do_start` -> `metaservice#start`->`recover()`.  we can delay this change, until we move recover out of recover. this change will bring a lot of changes in current metaservice UT, and so we better to do all the related works in a dedicated PR

Reply:
for now, we use `hs_utils::topological_sort` to both check circle and get the ordered_subtypes. if we call this in another function out of recovery, we need to pass the ordered_subtypes to `MetaBlkService::recover`, which need to change the signature of `MetaBlkService::recover` , and thus we need to do a lot of changes in the whole metaservice UT. can we do it in a seperate PR?

if we do not want to change the signature of `MetaBlkService::recover`, we can add ordered_subtypes as a member in MetaBlkService so that `MetaBlkService::recover` can read it, but it is generate from the `m_dep_topo_graph` and only be used once, i do not think it is a good fashion

Reply:
the sort in `recovery` or in `validate_register_dep` doesnt have any behavior change,  more for the code style. It looks to me also seems wired that `validate_register_dep` returns a order and passed to `MetaBlkService::recover`

Maybe do the topo-sort twice is a proper solution for the sake of code-style...

Honestly I dont think we need to spend more cycle on this style/preference and blocking the PR (other comments needed to be addressed),   as it is significant communication overhead to let other believe this is better compare to having another PR to fix the style to your prefer one.

Reply:
I don't get why we want to put more stuff in a function that is apparent not part of its job, and also would cause some confusion to the reader. The validation simply is NOT part of recovery, it is part of registration, if it fails, it shouldn't start recover process. Another benefit is if some use case want to call ::recover multiple times (either to measure the speed of recovery or test the dep/non-dep order correctness), it shouldn't do the sort multiple times, but just do recover with its clients.

Reply:
Logically it is ,  the point is the cycle-detection is actually a topo-sort,  so we either 

1. do the topo-sort twice , one for validation at the end of registration(validate_register_dep) , and one for recover.
2. let recovery() accept `ordered_subtypes` and `independent_subtypes`, which returned from `validate_register_dep` and passed in here.
3. Put the validation in recovery()

I think 2) will makes the flow more complicate and other caller of `recovery` will be  more confusing.   I am fine with 1 or 3. Though honestly dont see too much benefit.

Reply:
There is no need for independent_subtypes, where we can get from m_sub_info. 
This is not a major comment. Feel free to merge as it as. The rest looks good to me. 
But please expect some refactoring as to this part in next PR.